src/formatter.py

```diff
@@ -0,0 +1,125 @@
+
+
+class DictFormatter:
+    '''
+    Provides methods of serialisation of Game instance for various viewpoints
+    '''
+
+    def __init__(self, game):
+        self.game = game
+
+    def to_dict(self):
+        return DictFormatter.game_to_dict(self.game)
+
+    @staticmethod
+    def game_to_dict(game):
+        '''
+        :param game: Game object
+        :return: dict with (almost) complete state of the game. Decks return only the discard.
+        :players: dict of (id => Player dicts)
+            :name: str
+            :location: str
+            :action_count: int
+            :hand: dict of (id => Card dicts)
+                :name: str
+                :colour: str
+        :player_deck_discard: dict of (id => Card dicts)
+            :name: str
+            :colour: str
+        :infect_deck_discard: dict of (id => Card dicts)
+            :name: str
+            :colour: str
+        :diseases: dict of (colour => Disease dicts)
+            :colour: str
+            :cured: Boolean
+            :public_health: int
+        :cities: dict of (name => City dicts)
+            :name: str
+            :has_lab: Boolean
+            :colour: str
+            :infection_levels: dict (colour => lvl)
+        :infection_rate: int
+        :epidemic_count: int
+        '''
+        output = {}
+        output['players'] = {id: DictFormatter.player_to_dict(player) for id, player in enumerate(game.players)}
+        output['player_deck_discard'] = DictFormatter.deck_to_dict(game.player_deck)
+        output['infect_deck_discard'] = DictFormatter.deck_to_dict(game.infect_deck)
+        output['diseases'] = {colour: DictFormatter.disease_to_dict(disease) for colour, disease in
+                              game.diseases.items()}
+        output['cities'] = {name: DictFormatter.city_to_dict(city) for name, city in game.city_map.items()}
+        output['infection_rate'] = game.infection_rate
+        output['epidemic_count'] = game.epidemic_count
+
+        return output
+
+    @staticmethod
```

Comment:
Well, staticmethod can't be overridden. I did not expect that.

Reply:
Actually, it can. The one important thing is how you call it. If you expect static/class method can be overridden in a subclass, then it's better to call it via 'self' or 'cls' object.
In our previous case when using the 'staticmethod' decorator you loose information about the current class type, so you have to hardcode the current class name. In practice, that means, that if you want to, let's say, override card_to_dict to serialize just city name, you should also override all methods that invoke card_to_dict so they use overridden version.
When calling 'cls.card_to_dict', the 'cls' is not the class where the method defined, it's the type of object whose method is called.